### PR TITLE
gfx: restructure render-d3d12.cpp

### PR DESCRIPTION
**Pull request by csyonghe**
The existing gfx implementations follow a structural pattern that defines a `class *Device`, all implementation classes for other interfaces are nested inside the `*Device* class. The entire implementation for a target API is defined as gigantic class sitting in a single cpp file.

While defining everything inside the scope of a single parent class saves us from the trouble of separating out declaration and definition, it creates the problem that the implementation cannot be separated apart and must be ordered in a specific way so that some types are defined first to allow an implementation of another type to compile. It is also impossible to sparate some pieces of the implementation (e.g. shader-object related stuff) into a different file.

The complicated nesting structure also makes the code difficult to navigate and it can take a long time to find where a member is defined when intellisense is not working.

This changes restructures the d3d12 implementation to make it follow a more conventional C++ code structure: all classes are declared in a header file, then all implementations goes to a cpp file. No complicated type nesting. All d3d12 implementation code are placed under a `gfx::d3d12` namespace to avoid naming conflict with other targets.

Since this change touches every line of code in d3d12-render.cpp, I also ran clang-format on the entire file to make its formatting consistent.

### CODE REVIEW

File: tools/gfx/d3d11/render-d3d11.cpp [hunk 1]
```diff
@@ -2071,7 +2071,7 @@ SlangResult SLANG_MCALL createD3D11Device(const IDevice::Desc* desc, IDevice** o
     return SLANG_OK;
 }
 
-static void _initSrvDesc(IResource::Type resourceType, const ITextureResource::Desc& textureDesc, DXGI_FORMAT pixelFormat, D3D11_SHADER_RESOURCE_VIEW_DESC& descOut)
```
Comment (jsmall-nvidia) [inline tools/gfx/d3d11/render-d3d11.cpp:2074]: Nit pick: Is there a reason to what to make impls that are private to a cpp file, not use `_`? The `_` prefixing is of course not necessary, and is more typically used for non public methods, but here it seems appropriate.

File: source/core/slang-dictionary.h [hunk 2]
```diff
@@ -167,7 +167,7 @@ namespace Slang
 					return FindPositionResult(hashPos, -1);
 				}
 				numProbes++;
-				hashPos = (hashPos + GetProbeOffset(numProbes)) & bucketSizeMinusOne;
+				hashPos = (hashPos + GetProbeOffset(numProbes)) % bucketSizeMinusOne;
```
Comment (jsmall-nvidia) [inline source/core/slang-dictionary.h:170]: As discussed in other PR it seems this should either be `& bucketSizeMinusOne` (if bucketSize is pow 2) or `% (bucketSizeMinusOne + 1)`, if it's not.

**Review by jsmall-nvidia — COMMENTED** [review]
I'll wait til #2153 is merged so I can more easily see what the diffs are.

File: tools/gfx/d3d11/render-d3d11.cpp (hunk 1 above)
Comment (csyonghe) [reply tools/gfx/d3d11/render-d3d11.cpp:2074]: I intended to change the name in render-d3d12.cpp but somehow the refactoring feature mistakenly changes the name here. All these functions will eventually be moved out to separate files so perhaps it is fine to keep it as is.

File: tools/gfx/d3d12/render-d3d12.h [hunk 3]
```diff
@@ -1,11 +1,1799 @@
...
+
+// Define function pointer types for PIX library.
+typedef HRESULT(WINAPI* PFN_BeginEventOnCommandList)(
+    ID3D12GraphicsCommandList* commandList, UINT64 color, _In_ PCSTR formatString);
+typedef HRESULT(WINAPI* PFN_EndEventOnCommandList)(ID3D12GraphicsCommandList* commandList);
+
+static const Int kMaxNumRenderFrames = 4;
+static const Int kMaxRTVCount = 8;
+
+struct D3D12DeviceInfo
```
Comment (jsmall-nvidia) [inline tools/gfx/d3d12/render-d3d12.h:47]: Does it need D3D12 prefix when in the d3d12 namespace?

**Review by jsmall-nvidia — APPROVED** [review]
Why does the implementation of the methods move to the header? To be more like a regular C++ project - large method impls would typically remain in cpp file unless there is an important reason not to.

File: tools/gfx/d3d12/render-d3d12.h [hunk 4]
```diff
@@ -1,11 +1,1799 @@
...
+
+    virtual SLANG_NO_THROW Result SLANG_MCALL unmap(MemoryRange* writtenRange) override;
+
+    virtual SLANG_NO_THROW Result SLANG_MCALL setDebugName(const char* name) override;
+};
+
+class TextureResourceImpl : public TextureResource
+{
+public:
+    typedef TextureResource Parent;
```
[the remark this answers is not in the document]
Comment (csyonghe) [reply tools/gfx/d3d12/render-d3d12.h:314]: Yes, I noticed this as well. This is just original code and I will change it later.

File: tools/gfx/d3d12/render-d3d12.h (hunk 3 above)
Comment (csyonghe) [reply tools/gfx/d3d12/render-d3d12.h:47]: There is already a DeviceInfo type in `gfx` namespace, perhaps I should just rename it to something else.

**Pull request merged**